### PR TITLE
feat!: Change sums to be over TypeRows rather than Types

### PR DESCRIPTION
**Pull request by doug-q**

### CODE REVIEW

File: src/ops/controlflow.rs [hunk 1]
```diff
@@ -288,13 +285,3 @@ impl Case {
         &self.signature.output
     }
 }
-
-fn tuple_sum_first(tuple_sum_row: &TypeRow, rest: &TypeRow) -> TypeRow {
```
Comment (doug-q) [inline src/ops/controlflow.rs:292]: I have replaced uses with the new `TypeRow::extend`

File: src/types/check.rs [hunk 2]
```diff
@@ -68,7 +68,7 @@ fn type_sig_equal(v: &Hugr, t: &PolyFuncType) -> bool {
 }
 
 impl super::SumType {
-    /// Check that a [`Value`] is a valid instance of this [`SumType`].
+    /// Check that a [`Value`] is a valid instance of this [`crate::types::SumType`].
```
Comment (aborgna-q) [inline src/types/check.rs:71]:
```suggestion
    /// Check that a [`Value`] is a valid instance of this [`SumType`].
    ///
    ///   [`SumType`]: crate::types::SumType
```
So the link text is just the type

File: src/hugr/serialize.rs [hunk 3]
```diff
@@ -29,6 +29,7 @@ use super::{HugrError, HugrMut, HugrView};
 enum Versioned {
     /// Version 0 of the HUGR serialization format.
     V0(SerHugrV0),
+    V1(SerHugrV1),
```
Comment (cqc-alec) [inline src/hugr/serialize.rs:32]: Add docstring for V1?

File: src/hugr/serialize.rs [hunk 4]
```diff
@@ -57,6 +58,21 @@ struct SerHugrV0 {
     metadata: Vec<serde_json::Value>,
 }
 
+/// Version 1 of the HUGR serialization format.
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+struct SerHugrV1 {
```
Comment (cqc-alec) [inline src/hugr/serialize.rs:63]: I suppose we don't need `SerHugrV0` any more and can delete the struct definition?

File: src/types.rs [hunk 5]
```diff
@@ -215,10 +221,22 @@ impl TypeEnum {
 ///
 pub struct Type(TypeEnum, TypeBound);
 
+/// TODO
```
Comment (cqc-alec) [inline src/types.rs:224]: Add docstring.

File: src/types.rs [hunk 6]
```diff
@@ -215,10 +221,22 @@ impl TypeEnum {
...
+pub fn validate_each<'a>(
+    extension_registry: &ExtensionRegistry,
+    var_decls: &[TypeParam],
+    mut iter: impl Iterator<Item = &'a Type>,
+) -> Result<(), SignatureError> {
+    iter.try_for_each(|t| t.validate(extension_registry, var_decls))
+}
+
 impl Type {
+    /// TODO
```
Comment (cqc-alec) [inline src/types.rs:234]: Add docstring.

File: specification/schema/hugr_schema_v1.json [hunk 7]
```diff
@@ -0,0 +1,1618 @@
...
+                    "type": "array"
+                },
+                "sum_rows": {
+                    "items": {
+                        "items": {
+                            "$ref": "#/$defs/Type"
+                        },
+                        "type": "array"
+                    },
+                    "title": "Tuple Sum Rows",
```
Comment (cqc-alec) [inline specification/schema/hugr_schema_v1.json:373]:
```suggestion
                    "title": "Sum Rows",
```

File: specification/hugr.md [hunk 8]
```diff
@@ -426,8 +423,7 @@ output of each of these is a sum type, whose arity is the number of outgoing
 control edges; the remaining outputs are those that are passed to all
 succeeding nodes.
 
-The three nodes labelled "Const" are simply generating a TupleSum with one empty
-value to pass to the Output node.
+The three nodes labelled "Tag 0" are simply generating a 1-variant unary Sum (i.e. a Sum one variant with empty rows) to the Output node.
```
Comment (cqc-alec) [inline specification/hugr.md:426]:
```suggestion
The three nodes labelled "Tag 0" are simply generating a 1-variant unary Sum (i.e. a Sum of one variant with empty rows) to the Output node.
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]
Looks good! Just a few comments.

File: justfile [hunk 9]
```diff
@@ -8,7 +8,7 @@ test:
 
 # Auto-fix all clippy warnings
 fix:
-    cargo clippy --all-targets --all-features --workspace --fix --allow-staged
+    cargo clippy --all-targets --all-features --workspace --fix --allow-staged --allow-dirty
```
[the remark this answers is not in the document]
Comment (doug-q) [reply justfile:11]: This slipped in, but I do think it's an improvement (I usually have some untracked files). I will raise at standup and remove from here if there is less than unanimous agreement.

[the remark this answers is not in the document]
Comment (doug-q) [reply justfile]: Removed this

File: quantinuum-hugr/src/ops/constant.rs [hunk 10]
```diff
@@ -35,15 +35,17 @@ impl Const {
         &self.typ
     }
 
-    /// Sum of Tuples, used for branching.
-    /// Tuple rows are defined in order by input rows.
-    pub fn tuple_sum(
+    /// Creates a new Const Sum.  The value determined by `items` and is
```
Comment (cqc-alec) [inline quantinuum-hugr/src/ops/constant.rs:38]:
```suggestion
    /// Creates a new Const Sum.  The value is determined by `items` and is
```

File: specification/hugr.md [hunk 11]
```diff
@@ -428,8 +425,7 @@ output of each of these is a sum type, whose arity is the number of outgoing
 control edges; the remaining outputs are those that are passed to all
 succeeding nodes.
 
-The three nodes labelled "Const" are simply generating a TupleSum with one empty
-value to pass to the Output node.
+The three nodes labelled "Tag 0" are simply generating a 1-variant unary Sum (i.e. a Sum  of one variant with empty rows) to the Output node.
```
Comment (cqc-alec) [inline specification/hugr.md:428]:
```suggestion
The three nodes labelled "Tag 0" are simply generating a 1-variant unary Sum (i.e. a Sum of one variant with empty rows) to the Output node.
```

**Review by cqc-alec — APPROVED** [review]
Just a couple of pedantries.

File: quantinuum-hugr/src/ops/constant.rs [hunk 12]
```diff
@@ -35,15 +35,17 @@ impl Const {
...
+    /// type-checked `typ`
+    pub fn new_sum(
         tag: usize,
-        value: Value,
-        variant_rows: impl IntoIterator<Item = TypeRow>,
+        items: impl IntoIterator<Item = Const>,
+        typ: SumType,
     ) -> Result<Self, ConstTypeError> {
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply quantinuum-hugr/src/ops/constant.rs:44]: Yes, that makes sense.

**Pull request merged**